rclcpp/src/rclcpp/executor.cpp

```diff
@@ -699,9 +668,9 @@ Executor::wait_for_work(std::chrono::nanoseconds timeout)
           invalid_group_ptrs.push_back(weak_group_ptr);
           auto node_guard_pair = weak_nodes_to_guard_conditions_.find(weak_node_ptr);
           if (node_guard_pair != weak_nodes_to_guard_conditions_.end()) {
-            auto guard_condition = node_guard_pair->second;
+            const auto & guard_condition = node_guard_pair->second;
             weak_nodes_to_guard_conditions_.erase(weak_node_ptr);
-            memory_strategy_->remove_guard_condition(guard_condition);
+            memory_strategy_->remove_guard_condition(*guard_condition);
```

Comment:
Is it possible to have this be nullptr unexpectedly? Should we check it?

Reply:
Here we're in the situation in which a node could be expired: `weak_node_ptr.expired()`
If that the case the node's guard condition has been destructed, so the dereference `*guard_condition` is wrong since is now a dangling pointer. It could be `nullptr` or not, either way is dangerous.
The memory strategy should hold weak pointers of the guard condition, or have a map `{WeakNode, GuardCondition}` as the executor does, or share the same map. 
So the node's GuardCondition should be created as SharedPtr, and stored as WeakPtr in executor/memory_strategy. 
What do you think?

Reply:
What about storing only WeakNodes? I assume that regardless of the approach, before using the guard condition we will still have to check that the node has not gone out of scope, so at that point it may be worth directly calling `get_notify_guard_condition` every time

Reply:
Commit 9da50f1 fixes this problem (accessing an out-of-scope `GuardCondition` pointer) by restoring `remove_guard_condition` original way of taking directly a pointer to the guard condition to remove.
This way we also avoid to take the address of a `GuardCondition` inside the API, which might have gone out of scope.
I have rebased this PR with master. I think all the comments have been addressed now.